### PR TITLE
integrate ORT Downloader as a processor step

### PR DESCRIPTION
**Pull request by bs-ondem**
Resolves https://github.com/eclipse/antenna/issues/326

This PR implements a new processor step, in which the ORT Downloader is programmatically called. The ORT Downloader downloads the sources of the artifacts in a temporary directory if it has no local sources. Afterwards it will be archived to a ZIP file in the hardcoded directory `<AntennaTargetDirectory>/ort-downloader-result` and saved in the artifact model. 

New Dependencies: 
- com.github.heremaps.oss-review-toolkit:downloader:f91202964b

### Request Reviewer
@sschuberth @blaumeiser-at-bosch @neubs-bsi 

### Type of Change
new feature | documentation update

### How Has This Been Tested?
Run `example-projects/example-project` with `mvn clean package` and check the `ort-downloader-result` directory and the `sources.zip` file, which now contains also the sources of the JS components. 

### Checklist
- [x] All related issues are referenced in commit messages
- [x] I have updated the documentation accordingly to my changes

### CODE REVIEW

File: antenna-documentation/src/site/markdown/processors/ort-downloader.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+## ORT Downloader 
+The ORT Downloader processor downloads the sources of each artifact if no `ArtifactSourceFile` fact is present.
```
Comment (sschuberth) [inline antenna-documentation/src/site/markdown/processors/ort-downloader.md]: Just for my understanding: I assume that *if* `ArtifactSourceFile` is present, it points to a local existing file, correct? I.e., it does *not* point to a remote file location where the source code is hosted. Othwiser the sentence would not make sense to me, as I would expect the ORT Download to download remote files.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactVcsInfo.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+import org.eclipse.sw360.antenna.model.artifact.ArtifactFact;
+
+import java.util.Objects;
+import java.util.Optional;
+
+public class ArtifactVcsInfo implements ArtifactFact<ArtifactVcsInfo> {
+    private final VcsInfo vcsInfo;
+
+    public static class VcsInfo {
```
Comment (sschuberth) [inline core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactVcsInfo.java]: Could we simply directly use ORT's VcsInfo class here instead?

File: modules/ort/src/main/kotlin/resolver/OrtResultArtifactResolver.kt [hunk 3]
```diff
@@ -47,6 +47,11 @@ private fun mapSourceUrl(pkg: Package): ArtifactSourceUrl? =
         ArtifactSourceUrl(it)
     }
 
+private fun mapVcsInfo(pkg: Package) : ArtifactVcsInfo? =
+        pkg.vcs.takeUnless { it?.equals(VcsInfo.EMPTY) }?.let {
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/resolver/OrtResultArtifactResolver.kt]: Could can simplify `it?` to just `it` here as the result is non-nullable.

File: modules/ort/src/main/kotlin/utils/ArtifactToPackageMapper.kt [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                homepageUrl = "",
+                binaryArtifact = RemoteArtifact.EMPTY,
+                sourceArtifact = mapSourceRemoteArtifact(artifact) ?: RemoteArtifact.EMPTY,
+                vcs = mapVcsInfo(artifact) ?: VcsInfo.EMPTY
+                )
+    }
+
+    private fun mapIdentifier(artifact: Artifact) : Identifier? =
+        artifact.askFor(ArtifactCoordinates::class.java).takeIf { it.isPresent }?.let {
+            var coordinates = it.get()
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/utils/ArtifactToPackageMapper.kt]: This could be a `val`.

File: modules/ort/src/main/kotlin/utils/ArtifactToPackageMapper.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            val type = coordinates.mainCoordinate.type.toLowerCase()
+            val namespace = coordinates.mainCoordinate.namespace ?: ""
+            val name = coordinates.mainCoordinate.name ?: ""
+            val version = coordinates.mainCoordinate.version ?: ""
+            when (type) {
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/utils/ArtifactToPackageMapper.kt]: I'd propose to only map (Antenna) type to (ORT) type here for clarity, and construct the `Identifier` afterwards. Like:

            val ortType = when (type) {
                "maven" -> MAVEN_TYPE_KEY
                "npm" -> NPM_TYPE_KEY
                "nuget" -> NUGET_TYPE_KEY
                else -> UNMANAGED_TYPE_KEY
            }
            return Identifier(type = ortType, namespace = namespace, name = name, version = version)

File: modules/ort/src/main/kotlin/utils/ArtifactToPackageMapper.kt [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                "nuget" -> Identifier(type = NUGET_TYPE_KEY, namespace = namespace, name = name, version = version)
+                else -> {
+                    Identifier(type = UNMANAGED_TYPE_KEY, namespace = namespace, name = name, version = version)
+                }
+            }
+        }
+
+    private fun mapVcsInfo(artifact: Artifact) : VcsInfo? =
+        artifact.askFor(ArtifactVcsInfo::class.java).takeIf { it.isPresent }?.let {
+            var vcsInfo = it.get().vcsInfo
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/utils/ArtifactToPackageMapper.kt]: This could be a `val`.

File: modules/ort/src/main/kotlin/utils/ArtifactToPackageMapper.kt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            VcsInfo(type = VcsType(vcsInfo.type), url = vcsInfo.url, revision = vcsInfo.revision)
+        }
+
+    private fun mapSourceRemoteArtifact(artifact: Artifact) : RemoteArtifact? =
+        artifact.askFor(ArtifactSourceUrl::class.java).takeIf { it.isPresent }?.let {
+            RemoteArtifact.EMPTY.copy(url = it.get().get())
+        }
+
+
+}
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/utils/ArtifactToPackageMapper.kt]: Nit: Missing newline at end of file.

File: modules/ort/src/main/kotlin/utils/ArtifactToPackageMapper.kt [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    private fun mapVcsInfo(artifact: Artifact) : VcsInfo? =
+        artifact.askFor(ArtifactVcsInfo::class.java).takeIf { it.isPresent }?.let {
+            var vcsInfo = it.get().vcsInfo
+            VcsInfo(type = VcsType(vcsInfo.type), url = vcsInfo.url, revision = vcsInfo.revision)
+        }
+
+    private fun mapSourceRemoteArtifact(artifact: Artifact) : RemoteArtifact? =
+        artifact.askFor(ArtifactSourceUrl::class.java).takeIf { it.isPresent }?.let {
+            RemoteArtifact.EMPTY.copy(url = it.get().get())
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/utils/ArtifactToPackageMapper.kt]: Is the second `.get()` semantically the right thing to call here? Or should it better be `.toString()` to clarify we expect to get a string representation here?

File: modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    companion object {
+        const val ORT_DOWNLOADER_DIR = "ort-downloader-result"
+        private val LOGGER = LoggerFactory.getLogger(OrtDownloaderProcessor::class.java)
+    }
+
+    init {
+        workflowStepOrder = 350
+    }
+
+    private lateinit var ortDownloadFolder: File
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt]: I believe the variable name is misleading as this is not the directory ORT downloads to, but the directory you ZIP the sources to after ORT has downloaded them. Maybe better: `sourcesZipFolder`. As a side note, I personally prefer the term "directory" over "folder" as the latter is mostly used in the Windows world whereas the former is more neutral.

File: modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+    override fun configure(configMap: MutableMap<String, String>) {
+        ortDownloadFolder = context.toolConfiguration.antennaTargetDirectory.resolve(ORT_DOWNLOADER_DIR).toFile()
+        ortDownloadFolder.takeUnless { it.exists() }?.let { it.mkdirs() }
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt]: `exists()` is too unspecific here, it should better be `isDirectory`. Also, I find the logic a bit hard to follow as you're not using the non-null return value of `takeUnless` at all. How about simply:

    if (!ortDownloadFolder.isDirectory) ortDownloadFolder.mkdirs()

Or, yet better as you check also for an error (e.g. is already a file of the name name exists):

    if (!ortDownloadFolder.isDirectory && !ortDownloadFolder.mkdirs()) throw ...

File: modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+            it.sourceFile.isPresent
+        }.forEach {
+            val pkg = ArtifactToPackageMapper().apply(it)
+            val name = it.askFor(ArtifactCoordinates::class.java).takeIf { f -> f.isPresent }
+                    ?.let {
+                        it.get().mainCoordinate.canonicalize()
+                    } ?: it.askFor(ArtifactFilename::class.java).takeIf { f -> f.isPresent }
+                    ?.let { f -> f.get().bestFilenameEntryGuess.takeIf { fe -> fe.isPresent }
+                            ?.let { e -> e.get().filename }
+                    } ?: "unknown"
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt]: Here's a slightly reformatted variant of the above which also has a few redundant `let`s removed (simply apply the IDE inspections):

            val name = artifact.askFor(ArtifactCoordinates::class.java).takeIf {
                it.isPresent
            }?.get()?.mainCoordinate?.canonicalize() ?: artifact.askFor(ArtifactFilename::class.java).takeIf {
                it.isPresent
            }?.let { filename ->
                filename.get().bestFilenameEntryGuess.takeIf {
                    it.isPresent
                }?.get()?.filename
            } ?: "unknown"

File: modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+                    val zipFile = File(
+                            ortDownloadFolder,
+                            "${pkg.id.name.encodeOrUnknown()}-${pkg.id.version.encodeOrUnknown()}.zip"
+                    )
+                    LOGGER.debug("Pack source directory '${downloadResult.downloadDirectory.absolutePath}' to " +
+                            "'${zipFile.absolutePath}.")
+                    downloadResult.downloadDirectory.packZip(zipFile)
+
+                    if (zipFile.exists()) {
+                        it.addFact(ArtifactSourceFile(zipFile.toPath()))
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt]: `it` here also needs to be renamed to `artifact` if you apply my code above.

File: modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+            val name = it.askFor(ArtifactCoordinates::class.java).takeIf { f -> f.isPresent }
+                    ?.let {
+                        it.get().mainCoordinate.canonicalize()
+                    } ?: it.askFor(ArtifactFilename::class.java).takeIf { f -> f.isPresent }
+                    ?.let { f -> f.get().bestFilenameEntryGuess.takeIf { fe -> fe.isPresent }
+                            ?.let { e -> e.get().filename }
+                    } ?: "unknown"
+            try {
+                LOGGER.debug("Download sources via ORT Downloader for '${name}'")
+                val downloadResult = Downloader().download(pkg, createTempDir("ortDownloaderDirectory"))
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt]: It's probably good practice to explicitly delete temporary directories again after use, or at least when the JVM shuts down, e.g. by using

                val ortDownloadDirectory = createTempDir("ortDownloaderDirectory").also { it.deleteOnExit() }
                val downloadResult = Downloader().download(pkg, ortDownloadDirectory)

File: modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+                if (downloadResult.downloadDirectory.exists()) {
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt]: Better again: `isDirectory`.

File: modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+                    val zipFile = File(
+                            ortDownloadFolder,
+                            "${pkg.id.name.encodeOrUnknown()}-${pkg.id.version.encodeOrUnknown()}.zip"
+                    )
+                    LOGGER.debug("Pack source directory '${downloadResult.downloadDirectory.absolutePath}' to " +
+                            "'${zipFile.absolutePath}.")
+                    downloadResult.downloadDirectory.packZip(zipFile)
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt]: Note: This zips *all* sources of *all* components into a *single* ZIP file. I.e. you're not creating a ZIP per component. But I guess that is what you want, right?

File: modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+
+                    if (zipFile.exists()) {
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt]: Better again: `isFile`.

File: antenna-documentation/src/site/markdown/processors/ort-downloader.md (hunk 1 above)
Comment (bs-ondem) [reply antenna-documentation/src/site/markdown/processors/ort-downloader.md]: Correct `ArtifactSourceFile` points to a local existing file.d

File: modules/ort/src/main/kotlin/utils/ArtifactToPackageMapper.kt (hunk 8 above)
Comment (bs-ondem) [reply modules/ort/src/main/kotlin/utils/ArtifactToPackageMapper.kt]: The semantically correct usage is to call `askForGet` instead of `askFor`:
```
artifact.askForGet(ArtifactSourceUrl::class.java).takeIf { it.isPresent }?.let {
            RemoteArtifact.EMPTY.copy(url = it.get())
```
I will change it.

File: modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt (hunk 15 above)
Comment (bs-ondem) [reply modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt]: No I'm creating a ZIP per component, because `downloadResult.downloadDirectory` points to the  downloaded sources directory of the package.

File: core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactVcsInfo.java (hunk 2 above)
Comment (bs-ondem) [reply core/model/src/main/java/org/eclipse/sw360/antenna/model/artifact/facts/ArtifactVcsInfo.java]: If we use the ORT's VcsInfo class then the antenna-model has a direct dependency to ORT's model.

File: modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt [hunk 17]
```diff
@@ -33,39 +34,45 @@ class OrtDownloaderProcessor : AbstractProcessor() {
...
+                            sourcesZipDirectory,
                             "${pkg.id.name.encodeOrUnknown()}-${pkg.id.version.encodeOrUnknown()}.zip"
                     )
                     LOGGER.debug("Pack source directory '${downloadResult.downloadDirectory.absolutePath}' to " +
                             "'${zipFile.absolutePath}.")
                     downloadResult.downloadDirectory.packZip(zipFile)
 
-                    if (zipFile.exists()) {
-                        it.addFact(ArtifactSourceFile(zipFile.toPath()))
+                    if (zipFile.exists() && zipFile.isFile) {
```
Comment (sschuberth) [inline modules/ort/src/main/kotlin/workflow/processors/enricher/OrtDownloaderProcessor.kt]: `isFile` (and also `isDirectory`, BTW) imply `exists()`, so only either one is needed.

**Comment by sschuberth** [comment]
@bs-ondem, code-wise this looks good to me now, but Git-history-wise I'd expect you'd squash the fixup commits into the original commit before merging.
